scalding-core/src/main/scala/com/twitter/scalding/ExecutionOptimizationRules.scala

```diff
@@ -108,24 +109,346 @@ object ExecutionOptimizationRules {
         }
     }
 
-  object ZipWrite extends PartialRule[Execution] {
-    type InputType = (Config, Mode, Execution.Writer, ConcurrentExecutionContext)
-    type ResFun[T] =  InputType => Future[T]
+  sealed abstract class HList
+  object HList {
+    final case object HNil extends HList
+    final case class ::[A, B <: HList](head: A, tail: B) extends HList
 
-    case class NewFn[T, U](left: ResFun[T], right: ResFun[U]) extends (InputType => Future[(T, U)]) {
-      override def apply(tup: InputType): Future[(T, U)] =
-        Execution.failFastZip(left(tup), right(tup))(tup._4)
+    /*
+     * Some functions to deal with HLists
+     */
+    case class ConsFn[A, B <: HList](b: B) extends Function1[A, A :: B] {
+      def apply(a: A): A :: B = ::(a, b)
     }
+    case class Cons2Fn[A, B <: HList]() extends Function1[(A, B), A :: B] {
+      def apply(ab: (A, B)): A :: B = ::(ab._1, ab._2)
+    }
+    case class HeadFn[A]() extends Function1[A :: HNil.type, A] {
```

Comment:
this head would only work on a single element list right?

Reply:
yeah, I only use it that way, but I think it will be clearer actually if I make it generic on the tail too, and avoid the `HNil.type` which is a bit confusing.